libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1037 @@
+      else if (g_strcmp0 (interface_name, "com.endlessm.DiscoveryFeedWord") == 0)
+        g_ptr_array_add (word_proxies, ka_proxy);
+      else if (g_strcmp0 (interface_name, "com.endlessm.DiscoveryFeedQuote") == 0)
+        g_ptr_array_add (quote_proxies, ka_proxy);
+    }
+
+  zipped_word_quote_proxies = zip_two_ptr_arrays (word_proxies,
+                                                  quote_proxies,
+                                                  (PtrArrayZipElementFactory) word_quote_proxy_bundle_new,
+                                                  (GDestroyNotify) word_quote_proxy_bundle_free);
+
+  for (i = 0; i < zipped_word_quote_proxies->len; ++i)
+    {
+      WordQuoteProxyBundle *bundle = g_ptr_array_index (zipped_word_quote_proxies, i);
+      append_discovery_feed_word_quote_from_proxies (bundle->word_proxy,
+                                                     bundle->quote_proxy,
+                                                     cancellable,
+                                                     individual_task_result_completed,
+                                                     individual_task_result_closure_new (all_tasks_closure));
+    }
+}
+
+static void
+received_all_unordered_card_array_results_from_queries (GObject      *source,
+                                                        GAsyncResult *result,
+                                                        gpointer      user_data)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GPtrArray) results = g_task_propagate_pointer (G_TASK (result),
+                                                           &local_error);
+  g_autoptr(GPtrArray) result_arrays = g_ptr_array_new_with_free_func ((GDestroyNotify) g_ptr_array_unref);
+  g_autoptr(GPtrArray) flat_mapped_array = NULL;
+  g_autoptr(GTask) task = user_data;
+  guint i = 0;
+
+  /* This basically shouldn't happen, but handle it anyway */
+  if (results == NULL)
+    {
+      g_message ("Error getting all unordered card results: %s", local_error->message);
+      g_task_return_error (task, local_error);
+      return;
+    }
+
+  /* Go through each of the results, complain about the ones that failed
+   * but keep the ones that yielded some models. We'll flat-map them all
+   * together later */
+  for (i = 0; i < results->len; ++i)
+    {
+      g_autoptr(GPtrArray) result_array = g_task_propagate_pointer (g_ptr_array_index (results, i),
+                                                                    &local_error);
+
+      if (result_array == NULL)
+        {
+          g_message ("Query failed: %s", local_error->message);
+          g_clear_error (&local_error);
+          continue;
+        }
+
+      g_ptr_array_add (result_arrays, g_steal_pointer (&result_array));
+    }
+
+  flat_mapped_array = flat_map_ptr_arrays (result_arrays,
+                                           ref_object_copy_func,
```

Comment:
Wouldn't that pass more arguments on the stack than the function's calling convention would allow? (g_object_unref takes one arg, GCopyFunc is two args)